pkg/eks/addons/procedures.go

```diff
@@ -0,0 +1,160 @@
+}
+
+// Do implements the logic for the procedure
+func (p *DeleteAddonProcedure) Do(ctx context.Context) error {
+	input := &eks.DeleteAddonInput{
+		AddonName:   aws.String(p.name),
+		ClusterName: aws.String(p.plan.clusterName),
+	}
+
+	_, err := p.plan.eksClient.DeleteAddon(input)
+	if err != nil {
+		return fmt.Errorf("deleting eks addon %s: %w", p.name, err)
+	}
+
+	return nil
+}
+
+// Name is the name of the procedure
+func (p *DeleteAddonProcedure) Name() string {
+	return fmt.Sprintf("delete eks addon %s", p.name)
+}
+
+// UpdateAddonProcedure is a procedure that will update an EKS addon
+type UpdateAddonProcedure struct {
+	plan *addonsPlan
+	name string
+}
+
+// Do implements the logic for the procedure
+func (p *UpdateAddonProcedure) Do(ctx context.Context) error {
+	desired := p.plan.getDesired(p.name)
+
+	requestToken := string(uuid.NewUUID())
+	input := &eks.UpdateAddonInput{
+		AddonName:          &desired.Name,
+		AddonVersion:       &desired.Version,
+		ClusterName:        &p.plan.clusterName,
+		ResolveConflicts:   desired.ResolveConflict,
+		ClientRequestToken: &requestToken,
+	}
+	if desired.ServiceAccountRoleARN != "" {
+		input.ServiceAccountRoleArn = &desired.ServiceAccountRoleARN
+	}
+	_, err := p.plan.eksClient.UpdateAddon(input)
+	if err != nil {
+		return fmt.Errorf("updating eks addon %s: %w", p.name, err)
+	}
+
+	return nil
+}
+
+// Name is the name of the procedure
+func (p *UpdateAddonProcedure) Name() string {
+	return fmt.Sprintf("update eks addon %s", p.name)
+}
+
+// UpdateAddonTagsProcedure is a procedure that will update an EKS addon tags
+type UpdateAddonTagsProcedure struct {
+	plan *addonsPlan
+	name string
+}
+
+// Do implements the logic for the procedure
+func (p *UpdateAddonTagsProcedure) Do(ctx context.Context) error {
+	desired := p.plan.getDesired(p.name)
+	installed := p.plan.getInstalled(p.name)
+
+	input := &eks.TagResourceInput{
+		ResourceArn: &installed.ARN,
+		Tags:        convertTags(desired.Tags),
+	}
+	_, err := p.plan.eksClient.TagResource(input)
+	if err != nil {
+		return fmt.Errorf("updating eks addon tags %s: %w", p.name, err)
+	}
+
+	return nil
+}
+
+// Name is the name of the procedure
+func (p *UpdateAddonTagsProcedure) Name() string {
+	return fmt.Sprintf("update eks addon tags %s", p.name)
+}
+
+// CreateAddonProcedure is a procedure that will create an EKS addon for a cluster
+type CreateAddonProcedure struct {
+	plan *addonsPlan
+	name string
+}
+
+// Do implements the logic for the procedure
+func (p *CreateAddonProcedure) Do(ctx context.Context) error {
+	desired := p.plan.getDesired(p.name)
+
+	requestToken := string(uuid.NewUUID())
+	input := &eks.CreateAddonInput{
+		AddonName:          &desired.Name,
+		AddonVersion:       &desired.Version,
+		ClusterName:        &p.plan.clusterName,
+		ResolveConflicts:   desired.ResolveConflict,
+		Tags:               convertTags(desired.Tags),
+		ClientRequestToken: &requestToken,
+	}
+	if desired.ServiceAccountRoleARN != "" {
```

Comment:
It might be worth creating an issue to reexamine after some time the fact that it's relatively cumbersome to define these roles externally, especially because securing the role means setting a trust relationship that depends on the name/namespace of the service account, which is well known but not provided by the API.
At the moment I have to create a role beforehand, with knowledge of the needed policies and eventual metadata of the service account, before creating the addon.

Reply:
Yeah good thinking. Created #2219